homeassistant/components/google/__init__.py

```diff
@@ -188,7 +188,12 @@ def do_authentication(hass, hass_config, config):
 
     def step2_exchange(now):
         """Keep trying to validate the user_code until it expires."""
-        if now >= dt.as_local(dev_flow.user_code_expiry):
+
+        # oauth.step1_get_device_and_user_codes() returns a dev_flow.user_code_expiry
+        # timestamp that is in UTC, but with tzinfo == None.
+        user_code_expiry = dev_flow.user_code_expiry.replace(tzinfo=timezone.utc)
+
+        if now >= dt.as_local(user_code_expiry):
```

Comment:
Maybe we should use `track_utc_time_change` instead so we don't need to convert to local time?

Reply:
Indeed, sounds like that would make even more sense. I'll go ahead and rework the PR accordingly :+1:

Reply:
Doing just that doesn't seem to be enough though:

```
2021-08-20 21:56:30 ERROR (MainThread) [homeassistant] Error doing job: Future exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/srv/homeassistant_venv_3.8.11/lib/python3.8/site-packages/homeassistant/components/google/__init__.py", line 194, in step2_exchange
    if now >= dev_flow.user_code_expiry:
TypeError: can't compare offset-naive and offset-aware datetimes
```

```
diff --git homeassistant/components/google/__init__.py homeassistant/components/google/__init__.py
index 33afac6f57..5902353e8e 100644
--- homeassistant/components/google/__init__.py
+++ homeassistant/components/google/__init__.py
@@ -27,8 +27,8 @@ from homeassistant.const import (
 from homeassistant.helpers import discovery
 import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers.entity import generate_entity_id
-from homeassistant.helpers.event import track_time_change
-from homeassistant.util import convert, dt
+from homeassistant.helpers.event import track_utc_time_change
+from homeassistant.util import convert

 _LOGGER = logging.getLogger(__name__)

@@ -188,7 +188,7 @@ def do_authentication(hass, hass_config, config):

     def step2_exchange(now):
         """Keep trying to validate the user_code until it expires."""
-        if now >= dt.as_local(dev_flow.user_code_expiry):
+        if now >= dev_flow.user_code_expiry:
             hass.components.persistent_notification.create(
                 "Authentication code expired, please restart "
                 "Home-Assistant and try again",
@@ -216,7 +216,7 @@ def do_authentication(hass, hass_config, config):
             notification_id=NOTIFICATION_ID,
         )

-    listener = track_time_change(
+    listener = track_utc_time_change(
         hass, step2_exchange, second=range(0, 60, dev_flow.interval)
     )
```

So I guess we still need to do the `tzinfo` replace.

Reply:
Yes, that's still a good fix.

Reply:
Pushed a fixed version. Uses `track_utc_time_change` instead, but keeping the `.replace(tzinfo=timezone.utc)`.

Sorry if you wanted it as a separate commit, I'm so used to squashing and force pushing, so I did that before I realized that it might not be ok in this context.